mulog-filesystem-metrics/src/com/brunobonacci/mulog/publishers/filesystem_metrics.clj

```diff
@@ -0,0 +1,105 @@
+            [com.brunobonacci.mulog.buffer :as rb])
+  (:import  [java.nio.file FileSystems FileSystem FileStore]))
+
+
+
+(def store-hacks
+  [(when-let [klazz (try (Class/forName "sun.nio.fs.UnixFileStore")
+                         (catch ClassNotFoundException _ nil))]
+     (let [field (doto (.getDeclaredField ^Class klazz "file")
+                   (.setAccessible true))]
+       [klazz (fn unix-file-store-hack [store] (str (.get field store)))]))
+
+   (when-let [klazz (try (Class/forName "sun.nio.fs.WindowsFileStore")
+                         (catch ClassNotFoundException _ nil))]
+     (let [field (doto (.getDeclaredField ^Class klazz "root")
+                   (.setAccessible true))]
+       [klazz (fn windows-file-store-hack [store] (str (.get field store)))]))])
+
+
+
+(defn- file-store-path
+  [^FileStore store]
+  (->> store-hacks
+       (map (fn hack-path [[klazz hack]]
+              (when (and klazz (instance? klazz store))
+                (hack store))))
+       (remove nil?)
+       first))
+
+
+
+(s/def :fs/name string?)
+(s/def :fs/type string?)
+(s/def :fs/path (s/nilable string?))
+(s/def :fs/readonly? boolean?)
+(s/def :fs/total-bytes int?)
+(s/def :fs/unallocated-bytes int?)
+(s/def :fs/usable-bytes int?)
+(s/fdef capture-fs-metrics
+  :args (s/cat :store (partial instance? FileStore))
+  :ret (s/keys :req-un [:fs/name :fs/type :fs/path :fs/readonly?
+                        :fs/total-bytes :fs/unallocated-bytes :fs/usable-bytes]))
+
+
+
+(defn capture-fs-metrics [^FileStore store]
+  {:name (.name store)
+   :type (.type store)
+   :path (file-store-path store)
+   :readonly? (.isReadOnly store)
+   :total-bytes (.getTotalSpace store)
+   :unallocated-bytes (.getUnallocatedSpace store)
+   :usable-bytes (.getUsableSpace store)})
+
+
+
+(defn publish-fs-metrics
+  [^FileSystem fs filter]
+  (doseq [store (.getFileStores fs)
+          :let  [metrics (capture-fs-metrics store)]
+          :when (filter metrics)]
+    (u/log :mulog/filesystem-metrics-sampled
+      :filesystem-metrics metrics)))
+
+
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;                                                                            ;;
+;;                     ----==| P U B L I S H E R |==----                      ;;
+;;                                                                            ;;
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(deftype FilesystemMetricsPublisher [config buffer]
+
+  com.brunobonacci.mulog.publisher.PPublisher
+  (agent-buffer [_]
+    buffer)
+
+  (publish-delay [_]
+    (:sampling-interval config 60000))
+
+  (publish [_ buffer]
+    ;; sampling the file system metrics
+    (publish-fs-metrics (FileSystems/getDefault) (or (:disk-filter config) identity))))
+
+
+
+(def ^:const DEFAULT-CONFIG
+  {;; Interval in milliseconds between two samples
+   :sampling-interval 60000
+   :disk-filter nil})
```

Comment:
Instead of a *special* handling of `:disk-filter` I think it makes more sense to have the general [`:transform` support](https://cljdoc.org/d/com.brunobonacci/mulog/0.5.0/doc/how-to-write-custom-publishers#support-for-user-defined-transformations)

The `:transform` function is a function that takes a sequence of events and returns a presumably transformed sequence of events (`transform -> [event1 ... eventN] -> [event1' ... eventN']`).
Not only it can be used to filter unwanted events (same as `:disk-filter`), but it can be used to transform the events for the target system. For example, if for some reason you prefer sending the sizes in GB instead of bytes you could write a transformer function which does this before publishing.

I would suggest replacing the `:disk-filter` option the more general `:tranform`.

Reply:
Good idea, that's nicer. I've changed it.